src/models/src/entity/atproto.rs

```diff
@@ -0,0 +1,368 @@
+                            ));
+                        }
+
+                        if link.user_id != user.id {
+                            return Err(ErrorResponse::new(
+                                ErrorResponseType::BadRequest,
+                                "bad user_id in link cookie".to_string(),
+                            ));
+                        }
+
+                        if link.user_email != user.email {
+                            return Err(ErrorResponse::new(
+                                ErrorResponseType::BadRequest,
+                                "Invalid E-Mail".to_string(),
+                            ));
+                        }
+
+                        user.auth_provider_id = Some("atproto".to_owned());
+                        user.federation_uid = Some(claims_user_id.clone());
+
+                        Some(user)
+                    } else {
+                        return Err(ErrorResponse::new(ErrorResponseType::Forbidden, format!(
+                            "User with email '{}' already exists but is not linked to this provider.",
+                            user.email
+                        )));
+                    }
+                } else {
+                    None
+                }
+            }
+        };
+        debug!("user_opt:\n{:?}", user_opt);
+
+        let now = OffsetDateTime::now_utc().unix_timestamp();
+        let user = if let Some(mut user) = user_opt {
+            let mut old_email = None;
+            let mut forbidden_error = None;
+
+            if user.federation_uid.is_none()
+                || user.federation_uid.as_deref() != Some(&claims_user_id)
+            {
+                forbidden_error = Some("non-federated user or ID mismatch");
+            }
+
+            if user.auth_provider_id.as_deref() != Some("atproto") {
+                forbidden_error = Some("invalid login from wrong auth provider");
+            }
+
+            if let Some(err) = forbidden_error {
+                user.last_failed_login = Some(now);
+                user.failed_login_attempts =
+                    Some(user.failed_login_attempts.unwrap_or_default() + 1);
+                user.save(old_email).await?;
+
+                return Err(ErrorResponse::new(
+                    ErrorResponseType::Forbidden,
+                    err.to_string(),
+                ));
+            }
+
+            if Some(user.email.as_str()) != email.as_deref() {
+                old_email = Some(user.email);
+                user.email = email.as_ref().unwrap().to_string();
+            }
+
+            user.last_login = Some(now);
+            user.last_failed_login = None;
+            user.failed_login_attempts = None;
+
+            user.save(old_email).await?;
+            user
+        } else {
+            let new_user = User {
+                email: email.as_ref().unwrap().to_string(),
+                given_name: "N/A".to_string(),
```

Comment:
Can we provide something else / more meaningful than `N/A` for the `given_name` here?
The `N/A` was something I used for Github login, but it's not a nice solution, as it will throw an error in UI validation. I need to change it for Github as well so something like `Unknown`.